test/integration/integration_test.cc

```diff
@@ -2427,6 +2427,97 @@ TEST_P(IntegrationTest, RandomPreconnect) {
   }
 }
 
+/*
+This test ensures that the feature flag: envoy.restart_features.ensure_connection_retry
+enables retries on connection timeouts up to the limit per route instead of the limit
+coming from the downstream listener.
+*/
+TEST_P(IntegrationTest, EnsureConnectionRetry) {
+  // Set buffer limit upstream and downstream.
+  config_helper_.setBufferLimits(100, 64);
+  // Increase the base backoff to ensure we have time to see the connection failure and start
+  // the new upstream before the retry.
+  config_helper_.addRuntimeOverride("upstream.base_retry_backoff_ms", "200");
+  // enable feature flags envoy.restart_features.ensure_connection_retry
+  config_helper_.addRuntimeOverride("envoy.restart_features.ensure_connection_retry", "true");
+
+  // Start Envoy and close the socket of the fake upstream server.
+  initialize();
+  Network::Address::InstanceConstSharedPtr upstream_address = fake_upstreams_[0]->localAddress();
+  fake_upstreams_[0]->closeSocket();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto encoder_decoder = codec_client_->startRequest(
+      Http::TestRequestHeaderMapImpl{{":method", "POST"},
+                                     {":path", "/test/long/url"},
+                                     {":scheme", "http"},
+                                     {":authority", "sni.lyft.com"},
+                                     {"x-forwarded-for", "10.0.0.1"},
+                                     {"x-envoy-retry-on", "connect-failure"},
+                                     {"x-envoy-max-retries", "5"}});
+
+  auto& encoder = encoder_decoder.first;
+  auto& response = encoder_decoder.second;
+  // Send more data than the buffer limit.
+  codec_client_->sendData(encoder, 256, true);
```

Comment:
Do you also want to simulate multiple read situations?

Reply:
For multiple reads, it means the data from the buffer can be through multiple reads from I/O and then the buffer limit can be triggered. Did you try this scenario?